### PR TITLE
Fix spectator mode not showing when a spectated user quits correctly

### LINKED ISSUE #25535: Quitting map leaves spectator in gameplay screen
Seems pretty bad.

### Discussed in https://github.com/ppy/osu/discussions/25531

<div type='discussions-op-text'>

<sup>Originally posted by **uzervlad** November 22, 2023</sup>
After quitting gameplay, spectator is left on the gameplay screen

https://github.com/ppy/osu/assets/24654479/7666ff9a-52a5-4c9c-822e-4010358e0e64

</div>

### PR DESCRIPTION
**Pull request by peppy**
This PR switches `SoloSpectatorScreen` to use a nested screen for gameplay.

Spectator relies on the `SpectatorScreen` (which importantly, is not the `SpectatorPlayer` but a separate screen) to be alive and present in order to function correctly, due to `Schedule`d operations).

This was already true for multiplayer but not for solo play, so it seems logical to standardise things here.

Closes https://github.com/ppy/osu/issues/25535.

---

https://github.com/ppy/osu/assets/191335/f6ecdc22-7c48-454b-9d63-145f1b37b546

### CODE REVIEW

File: osu.Game/Screens/Play/SoloSpectatorScreen.cs [hunk 1]
```diff
@@ -218,7 +224,47 @@ void start()
...
+                    {
+                        newStack = new OsuScreenStack()
+                    }
+                }, container =>
+                {
+                    screenStackContainer?.FadeOut(200);
+                    screenStackContainer?.Expire();
+
+                    AddInternal(screenStackContainer = container);
+                    playerScreenStack = newStack;
```
Comment (bdach) [inline osu.Game/Screens/Play/SoloSpectatorScreen.cs:249]: I can't really tell if this is safe or not. Especially given this line:

https://github.com/ppy/osu/blob/bce32103ee8a7fe08f66f848052bc91e45ee17db/osu.Game/Screens/Play/SoloSpectatorScreen.cs#L212

Like, what is the guarantee that the following doesn't happen?

- `LoadComponentAsync()` called the first time => `newStack` is assigned the first time
- `LoadComponentAsync()` called the second time => `newStack` is assigned the second time
- second async load completes
- first async load completes => screen stack container from second load is expired
- at which point `playerScreenStack = newStack` happens => playerScreenStack is a child of an expired container

File: osu.Game/Online/Spectator/SpectatorClient.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ public abstract partial class SpectatorClient : Component, ISpectatorClient
         /// Whether the <see cref="SpectatorClient"/> is currently connected.
         /// This is NOT thread safe and usage should be scheduled.
         /// </summary>
-        public abstract IBindable<bool> IsConnected { get; }
+        protected abstract IBindable<bool> IsConnected { get; }
```
Comment (bdach) [inline osu.Game/Online/Spectator/SpectatorClient.cs:36]: Any particular reason for encapsulating this? I rely on this being public in https://github.com/ppy/osu/pull/25522.

Comment (peppy) [reply osu.Game/Online/Spectator/SpectatorClient.cs:36]: No reason, just wasn't being used.

Comment (bdach) [reply osu.Game/Online/Spectator/SpectatorClient.cs:36]: I guess I can just re-expose it again in mine when resolving merge conflicts.

**Comment by bdach** [comment]
Transition on retry is a tad broken:

https://github.com/ppy/osu/assets/20418176/16fa1bb2-2df2-426a-905d-7af88bf78f88

osu! logo should not be hanging around like that there.

**Pull request closed without merging**